lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1363,41 +1366,88 @@ static void diagnoseImplicitSelfUseInClosure(TypeChecker &TC, const Expr *E,
-        if (isImplicitSelfUse(DSCE->getBase()) &&
+        if (isImplicitSelfParamUse(DSCE->getBase()) &&
             isa<DeclRefExpr>(DSCE->getFn())) {
           auto MethodExpr = cast<DeclRefExpr>(DSCE->getFn());
+          memberLoc = DSCE->getLoc();
           TC.diagnose(DSCE->getLoc(),
                       diag::method_call_in_closure_without_explicit_self,
-                      MethodExpr->getDecl()->getBaseName().getIdentifier())
-            .fixItInsert(DSCE->getLoc(), "self.");
-          return { false, E };
+                      MethodExpr->getDecl()->getBaseName().getIdentifier());
         }
-
+      
+      // This error can be fixed by either capturing self explicitly (if in an
+      // explicit closure), or referencing self explicitly.
+      if (memberLoc.isValid()) {
+        if (auto *CE = dyn_cast<ClosureExpr>(ACE)) {
+          // If we've already captured something with the name "self" other than
+          // the actual self param, offer special diagnostics.
+          if (auto *VD = CE->getCapturedSelfDecl()) {
+            // Either this is a weak capture of self...
+            if (VD->getType()->is<WeakStorageType>()) {
+              TC.diagnose(VD->getLoc(), diag::note_self_captured_weakly);
+            // ...or something completely different.
+            } else {
+              TC.diagnose(VD->getLoc(), diag::note_other_self_capture);
+            }
+            // Bail on the rest of the diagnostics. Offering the option to
+            // capture 'self' explicitly will result in an error, and using
+            // 'self.' explicitly will be accessing something other than the
+            // self param.
+            // FIXME: We could offer a special fixit in the [weak self] case to insert 'self?.'...
+            return { false, E };
+          }
+          
+          auto diag = TC.diagnose(CE->getLoc(),
+                                  diag::note_capture_self_explicitly);
+          // If there's a token immediately following the opening brace of the
+          // closure, we may need to pad the fix-it with a space.
+          auto nextLoc = CE->getLoc().getAdvancedLoc(1);
+          auto next = Lexer::getTokenAtLocation(TC.Context.SourceMgr,
+                                                CE->getLoc().getAdvancedLoc(1));
+          std::string trailing = next.getLoc() == nextLoc ? " " : "";
+          if (CE->getInLoc().isInvalid())
+            diag.fixItInsertAfter(CE->getLoc(), " [self] in" + trailing);
+          else if (CE->getBracketRange().isInvalid())
+            diag.fixItInsertAfter(CE->getLoc(), " [self]" + trailing);
+          else {
+            // Insert into an existing capture list
+            auto range = CE->getBracketRange();
+            if (range.Start.getAdvancedLoc(1) == range.End)
+              diag.fixItInsertAfter(range.Start, "self");
+            else
+              diag.fixItInsertAfter(range.Start, "self, ");
```

Comment:
Oh neat, didn't realize that `fixItInsertAfter` could be used like that--that's much nicer. The `trailing` logic in lines (1425-1430) should handle the case of a token immediately following the opening brace, but let me know if I'm missing an edge case there.

You're right about the `getAdvancedLoc(1)` logic there. Will have to do a little more work to handle that case.

Reply:
Well, I'm not 100% sure how it will work out. 😅 If it does work, I think it would be easier to understand. (Maybe it'd be even easier to understand if you had named variables for the different conditions, like `needsBrackets`, `needsComma`, `needsIn`, etc.)

Reply:
Seems like breaking up the `fixItInsertAfter` calls unfortunately offers separate fix-its :( will refactor this to make the logic a bit clearer though.